homeassistant/components/tfl/config_flow.py

```diff
@@ -0,0 +1,266 @@
+    {
+        vol.Required(CONF_STOP_POINTS): TextSelector(TextSelectorConfig(multiple=True)),
+    }
+)
+
+
+@dataclass
+class ValidationResult:
+    """Define a validation result."""
+
+    errors: dict[str, str] = field(default_factory=dict)
+    description_placeholders: dict[str, str] = field(default_factory=dict)
+
+
+async def validate_app_key(hass: HomeAssistant, app_key: str) -> ValidationResult:
+    """Validate the user input for app_key."""
+
+    errors: dict[str, str] = {}
+
+    _LOGGER.debug("Validating app_key")
+    try:
+        stop_point_api = stopPoint(app_key)
+        # Make a random, cheap, call to the API to validate the app_key
+        categories = await call_tfl_api(hass, stop_point_api.getCategories)
+        _LOGGER.debug("Validating app_key, got categories=%s", categories)
+    except CannotConnect:
+        errors["base"] = "cannot_connect"
+    except InvalidAuth:
+        errors["base"] = "invalid_auth"
+    except Exception:  # pylint: disable=broad-except
+        _LOGGER.exception("Unexpected exception")
+        errors["base"] = "unknown"
+
+    return ValidationResult(errors)
+
+
+async def validate_stop_point(
+    hass: HomeAssistant, app_key: str, stop_point: str
+) -> ValidationResult:
+    """Validate the user input for stop point."""
+    errors: dict[str, str] = {}
+    description_placeholders: dict[str, str] = {}
+
+    _LOGGER.debug("Validating stop_point=%s", stop_point)
+
+    try:
+        stop_point_api = stopPoint(app_key)
+        _LOGGER.debug("Validating stop_point=%s", stop_point)
+        arrivals = await call_tfl_api(
+            hass, stop_point_api.getStationArrivals, stop_point
+        )
+        _LOGGER.debug("Got for stop_point=%s, arrivals=%s", stop_point, arrivals)
+    except HTTPError as exception:
+        if exception.code == 404:
+            errors["base"] = "invalid_stop_point"
+            description_placeholders["stop_point"] = stop_point
+        else:
+            errors["base"] = "unknown"
+    except CannotConnect:
+        errors["base"] = "cannot_connect"
+    except InvalidAuth:
+        errors["base"] = "invalid_auth"
+    except Exception:  # pylint: disable=broad-except
+        _LOGGER.exception("Unexpected exception")
+        errors["base"] = "unknown"
+
+    return ValidationResult(errors, description_placeholders)
+
+
+async def validate_stop_points(
+    hass: HomeAssistant, app_key: str, stop_points: list[str]
+) -> ValidationResult:
+    """Validate the stop points."""
+
+    for stop_point in stop_points:
+        validation_result = await validate_stop_point(
+            hass,
+            app_key,
+            stop_point,
+        )
+        if validation_result.errors:
+            break
+
+    return validation_result
+
+
+class TfLConfigFlow(ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for Transport for London."""
+
+    VERSION = 1
+    data: dict[str, Any] = {}
+    options: dict[str, Any] = {}
+
```

Comment:
Don't create mutable objects like dicts or lists on class level, this means they will be shared between all instances of that class. Rather create them in the constructor so they are created on instance level

Reply:
Doh, of course. Thanks. I'm a Java / Kotlin dev by trade so whilst this is obvious I will blame it on my novice python skills even though the principles are the same ;-)

Reply:
Fixed in 60bce47a177403855332c56c0b5958470540defe